python/cudf/cudf/core/dataframe.py

```diff
@@ -3421,11 +3421,10 @@ def as_gpu_matrix(self, columns=None, order="F"):
         ):
             raise TypeError("non-numeric data not yet supported")
 
-        if all([is_datetime_dtype(col.dtype) for col in cols]):
-            # np bug: https://github.com/numpy/numpy/issues/17428
-            dtype = np.dtype("datetime64[ms]")
-        else:
-            dtype = np.find_common_type(cols, [])
+        try:
+            dtype = np.result_type(*cols)
+        except ValueError:
```

Comment:
Sorry, random fly-by comment. This mainly be a TypeError, or maybe both?  And I noticed one strange current behaviour in NumPy:
```
In [2]: np.result_type("M8", "m8")                                                                                      
Out[2]: dtype('<M8')
```
which doesn't make sense, but I wouldn't be surprised if we rely on it elsewhere :(. (I think this might also happen for integers and timedelta, which is pretty terrible as well, and now I wonder if also pandas has to cook its own soup here to get it right, you may want to ping them?).

Reply:
Thanks for the catch, there maybe both:
```
>>> import numpy as np
>>> np.result_type(np.int64, np.datetime64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 6, in result_type
TypeError: invalid type promotion
>>> np.result_type()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 6, in result_type
ValueError: at least one array or dtype is required
```
Will add to next commit to catch for both.

`np.promote_type` also has the same issue:
```
>>> np.promote_types("<M8", "<m8")
dtype('<M8')
```

We should probably make up some tests for `np.result_type` before using it.
cc @brandon-b-miller for dtype related issue.

Reply:
> Sorry, random fly-by comment. This mainly be a TypeError, or maybe both? And I noticed one strange current behaviour in NumPy:
> 
> ```
> In [2]: np.result_type("M8", "m8")                                                                                      
> Out[2]: dtype('<M8')
> ```
> 
> which doesn't make sense, but I wouldn't be surprised if we rely on it elsewhere :(. (I think this might also happen for integers and timedelta, which is pretty terrible as well, and now I wonder if also pandas has to cook its own soup here to get it right, you may want to ping them?).

Should I raise another issue on numpy issue tracker? If not already done so.

Reply:
On a separate thought - `np.result_type` did not specify the kind of operation being taken, so if we construct the following op:
```
def op(datetime, timedelta):
    return datetime + timedelta
```
The result_type is expected to be `M8`

Reply:
Yes, which is why the behaviour is probably there, Pandas calls this "concatenate type" I think. I call it "common dtype" in the NumPy restructure instead of `result_type`. Since `result_type` sounds like the promotion in functions, but for that you need to know the function (such as addition) as you point out.  So "concatenate dtype" is a pretty good name, because it is the correct dtype to use if you connect both things.

I know about these issues, but I doubt there is an open issue about it.  I am trying to focus on restructuring the internals rather than these details, but they are very important, so I think opening an issue would make sense... These are things that should have been fixed years ago, but e.g. pandas probably opted to implement their own version instead because it is faster/quicker (which is true).

Reply:
Are there actually any ops between `M` and `m` that *dont* lead to `M`? Here's what I found at least:

np.add.types and np.subtract.types contain `Mm->M`, `mM->M` and `mm->m` but nothing else
np.multiply, np.true_divide, np.floor_divide and np.power don't allow for `Mm` or `mM` at all as far as I can tell
np.remainder has a loop for `mm->m` but not `mM` or `Mm` 

so as far as I can tell any time `Mm` or `mM` is used, we always get `M`. From that perspective `np.result_type` makes sense, unless I am mistaken.

Reply:
@brandon-b-miller can't think of any "math" or ufunc-like either. But you have the issue here of e.g. transposing a dataframe or `np.concatenate([DatetimeArray, TimedeltaArray])`. The concatenate should obviously fail (requiring going to `object` if desired).  I have not tried to check what this PR is about, so I guess it likely does not matter here.

So I consider this a bug, even though it pans out for NumPy: It will try to concatenate it, but that requires casting and the cast will end up failing.  And that bug means that pandas must have its own `result_type` (I am sure it does, but I don't know exactly where or what).

Reply:
Looks like Pandas uses this: https://github.com/pandas-dev/pandas/blob/c04e2316c5f3d65c55a83ad3a0eb7092b12d1181/pandas/core/dtypes/cast.py#L1447-L1499

Which uses `np.find_common_type` with some extra logic / fallbacks.